### PR TITLE
Refactor to conform with ES-Shim API Contract

### PR DESCRIPTION
**Pull request by JakeChampion**
With the two together, it is impossible to test whether the module
works in engines which partially support Array.prototype.find.

With the two separated, we can test in all engines and we can more
easily use this polyfill inside of the [Financial Times's polyfill-service](https://github.com/Financial-Times/polyfill-service).

### CODE REVIEW

**Comment by JakeChampion** [comment]
CI doesn't seem to be set-up for this repository anymore, I have ran the tests in multiple versions of NodeJS myself (Versions 0.10/0.11/0.12/4/5). [The results are in a gist](https://gist.github.com/JakeChampion/e330205c7bb968a6cf5f). All the tests are passing.

File: array-prototype-find.js [hunk 1]
```diff
@@ -1,9 +1,7 @@
 // Array.prototype.find - MIT License (c) 2013 Paul Miller <http://paulmillr.com>
 // For all details and docs: https://github.com/paulmillr/array.prototype.find
 (function(globals){
-  if (Array.prototype.find) return;
-
-  var find = function(predicate) {
+  Array.prototype.find = function find (predicate) {
```
Comment (ljharb) [inline array-prototype-find.js]: it'd be nice to change this file so the implementation could be used without the shimming behavior - ie, move the shim assignment into `detect.js`?

Comment (JakeChampion) [inline array-prototype-find.js]: Hmm, I'm wanting to get the shimming behaviour without the detect. What do you think about having three files? `detect.js`, `polyfill.js`, `shim.js`?

Comment (ljharb) [inline array-prototype-find.js]: yes, let's please do that. see also: https://github.com/es-shims/es-shim-api

Comment (JakeChampion) [inline array-prototype-find.js]: That looks like a good specification to follow. Thanks for the link.

**Comment by JakeChampion** [comment]
@ljharb I've modified the project structure to now follow the ES-Shim API Contract. Hopefully I've followed it correctly.

File: index.js [hunk 2]
```diff
@@ -1,32 +1,22 @@
...
+	ES.RequireObjectCoercible(array);
+	return implementation.apply(array, slice.call(arguments, 1));
+};
+define(boundFindShim, {
+	implementation: implementation,
+	getPolyfill: getPolyfill,
+	shim: shim
+});
+
+module.exports = boundFindShim;
```
Comment (ljharb) [inline index.js]: trailing newline (you may want to configure your editor :-))

File: polyfill.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+module.exports = function getPolyfill() {
+	return Array.prototype.find || require('./implementation');
```
Comment (ljharb) [inline polyfill.js]: There is a bug in early implementations of `Array#find` that this should be detecting - https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L1432-L1440

File: shim.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+'use strict';
+
+var define = require('define-properties');
+var getPolyfill = require('./polyfill');
+
+module.exports = function shimArrayPrototypeFind() {
+	var polyfill = getPolyfill();
+	if (Array.prototype.find !== polyfill) {
+		define(Array.prototype, { find: polyfill });
```
Comment (ljharb) [inline shim.js]: `define` won't override existing properties unless there's a corresponding "should override" function - ie, here you want `define(Array.prototype, { find: polyfill }, { find: function () { return true; })`. Alternatively you could move the "if" condition into the predicate function.

**Comment by ljharb** [comment]
This looks great!

@paulmillr will have to enable travis-ci, however, since non-organization-owned repos can't enable admin  access for others.

File: index.js (hunk 2 above)
Comment (JakeChampion) [inline index.js]: I don't have trailing new lines on by default however, my editor will follow a project's style guide via [editorconfig](http://editorconfig.org/) and [eslint](http://eslint.org/). I can submit a PR with the styles you are wanting me to follow if you want?

File: polyfill.js (hunk 3 above)
Comment (JakeChampion) [inline polyfill.js]: Good catch!

File: shim.js (hunk 4 above)
Comment (JakeChampion) [inline shim.js]: Thanks for spotting this :-)

File: index.js (hunk 2 above)
Comment (ljharb) [inline index.js]: sure, an editorconfig wouldn't hurt

**Comment by ljharb** [comment]
It would also be ideal to add `@es-shims/api` as a dev dependency, and use it to run basic conformance tests - like https://github.com/es-shims/String.prototype.trim/blob/master/package.json#L20 for example.

**Comment by JakeChampion** [comment]
@ljharb Comments have been addressed :-)

File: implementation.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+// Array.prototype.find - MIT License (c) 2013 Paul Miller <http://paulmillr.com>
+// For all details and docs: https://github.com/paulmillr/array.prototype.find
+'use strict';
+var ES = require('es-abstract/es6');
+
+module.exports = function find (predicate) {
```
Comment (ljharb) [inline implementation.js]: please remove the space between the function's name and the argument list

**Comment by ljharb** [comment]
Previously, this project used 2-space indentation - most of your added files use tabs. I prefer tabs, so either one is fine, but we should make sure the whole project is consistent one way or another.

**Comment by JakeChampion** [comment]
@ljharb Migrated to tabs, added the change to `.editorconfig` for JS and JSON files. I've also removed the space between the function name and argument list. May I ask if you are following a certain community style guide? I could include it in a subsequent pull request if you think that would be helpful for contributors.

File: implementation.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// Array.prototype.find - MIT License (c) 2013 Paul Miller <http://paulmillr.com>
+// For all details and docs: https://github.com/paulmillr/array.prototype.find
+'use strict';
+var ES = require('es-abstract/es6');
+
+module.exports = function find(predicate) {
+  var list = ES.ToObject(this);
```
Comment (ljharb) [inline implementation.js]: this file still uses spaces

File: tests/test.js [hunk 7]
```diff
@@ -81,11 +80,32 @@ var runTests = function () {
...
+    Object.prototype[1] = 42;
+    runTests(find.implementation);
+    delete Object.prototype[1];
+  });
+});
+
+describe('shim', function() {
+	find.shim();
+	var implementation = Array.prototype.find;
+	console.log(implementation);
```
Comment (ljharb) [inline tests/test.js]: remove this?

File: tests/test.js [hunk 8]
```diff
@@ -81,11 +80,32 @@ var runTests = function () {
...
\ No newline at end of file
+  describe('clean Object.prototype', function() {
+    runTests(implementation);
+  });
+
+  describe('polluted Object.prototype', function() {
+    Object.prototype[1] = 42;
+    runTests(implementation);
+    delete Object.prototype[1];
+  });
```
Comment (ljharb) [inline tests/test.js]: still some spaces here too

**Comment by ljharb** [comment]
@JakeChampion somewhere between https://github.com/ljharb/eslint-config and https://github.com/airbnb/javascript - but "no space after the function name" and "always add a trailing newline" is pretty universal afaict.

**Comment by JakeChampion** [comment]
@ljharb Updated

**Comment by ljharb** [comment]
This LGTM - I'm going to give it a few days before merging though.

**Comment by JakeChampion** [comment]
All good with me, there's no rush with this work :-)

I'm happy to squash the commits if you would like a cleaner history

**Comment by ljharb** [comment]
Absolutely that would be appreciated - not down to 1, of course, but combining all the conceptually related ones so there's a smaller number of conceptually atomic commits would be great :-)

**Comment by JakeChampion** [comment]
I've squashed conceptually related commits together. If you think more can be squashed together I will happily do them also.

**Comment by paulmillr** [comment]
Thanks Jake.

**Pull request merged**